src/targets/gpu/gemm_impl.cpp

```diff
@@ -95,7 +95,8 @@ void gemm_impl(context& ctx,
                T alpha,
                T beta,
                bool int8_x4_format,
-               bool compute_fp32)
+               bool compute_fp32,
+               bool can_unbatch)
```

Comment:
There is no need to pass in a flag, you can just check that `num_matrices > 0 and get_batch_stride(batch_args[1]) == 0`, right?

Reply:
At least in the case of the GEMMs this applies to in huggingface bert, the B matrices are all literals. After propagate_constants they're no longer broadcasted because they get replaced with a literal of the broadcasted shape. I wasn't able to figure out a way to tell if they were originally broadcasted from the state of the graph at gemm_impl, so I moved upstream to lowering, but that is still after propagate_constants, so I had to use the slightly less clear logic that you see there. 
I also wrote a version where we run a pass between simplify_algebra and propagate_constants to insert a reshape on A and skip the broadcast on B, but PR #1280 makes that less straight forward since B comes from another dot operator until after propagate_constants.

Reply:
> At least in the case of the GEMMs this applies to in huggingface bert, the B matrices are all literals. After propagate_constants they're no longer broadcasted because they get replaced with a literal of the broadcasted shape.

We shouldn't be turning broadcast literals into literals in propagate_constants(unless there is a bug). We do turn broadcasted literal into literals in eliminate_contiguous here:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/eliminate_contiguous.cpp#L137

Which happens when `compute_shape` throws an error without contiguous. This most likely comes from here:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/targets/gpu/include/migraphx/gpu/gemm.hpp#L77

Where we reject broadcasted inputs. However, in #1280 I remove this check, so it should remain a broadcasted input.

Reply:
I was just looking back through it, and I was remembering a couple of things incorrectly. So for 3/4 of them, the B input does remain broadcasted at the lowering pass, so I can update that to check that the batch dim is broadcasted and remove the constraint that they're constants. 
The case I was thinking of is just the first GEMM, the horizontally fused one. Before propagate_constant: 
```
main:@51 = @literal{ ... } -> half_type, {768, 768}, {768, 1}
main:@52 = multibroadcast[out_lens={2, 768, 768}](main:@51) -> half_type, {2, 768, 768}, {0, 768, 1}
main:@53 = contiguous(main:@52) -> half_type, {2, 768, 768}, {589824, 768, 1}
main:@54 = @literal{ ... } -> half_type, {768, 768}, {768, 1}
main:@55 = multibroadcast[out_lens={2, 768, 768}](main:@54) -> half_type, {2, 768, 768}, {0, 768, 1}
main:@56 = contiguous(main:@55) -> half_type, {2, 768, 768}, {589824, 768, 1}
main:@57 = @literal{ ... } -> half_type, {768, 768}, {768, 1}
main:@58 = multibroadcast[out_lens={2, 768, 768}](main:@57) -> half_type, {2, 768, 768}, {0, 768, 1}
main:@59 = contiguous(main:@58) -> half_type, {2, 768, 768}, {589824, 768, 1}
main:@60 = concat[axis=2](main:@53,main:@56,main:@59) -> half_type, {2, 768, 2304}, {1769472, 2304, 1}
main:@61 = add(main:@48,main:@50) -> half_type, {2, 16, 768}, {12288, 768, 1}
main:@72 = dot(main:@61,main:@60) -> half_type, {2, 16, 2304}, {36864, 2304, 1}
```
And at the lowering pass:
```
main:@0 = @literal{ ... } -> half_type, {2, 768, 2304}, {1769472, 2304, 1}
main:@34 = gpu::precompile_op[op=pointwise](main:@29,main:@32,main:@33), [main:pointwise2] -> half_type, {2, 16, 768}, {12288, 768, 1}
main:@35 = dot(main:@34,main:@0) -> half_type, {2, 16, 2304}, {36864, 2304, 1}
```
So I think for that one, it's the concat that causes it to get replaced. 

I'll base a branch off of #1280 and see how I can improve it from there.

Reply:
We can update simplify_reshapes to find a concat with broadcasted inputs to turn into a concat first and then a broadcast as long as the concat axis is not being broadcasted.

Reply:
Sounds good. The other three can be checked from gemm_impl after merging in 1280, so I'll make that change to simplify_algebra and then we should be able to remove the flag that I had added.

Reply:
Well I said simplify_reshapes not simplify_algebra. We have similar find_concat_transpose in simplify_reshapes.

Reply:
Sorry, typed that incorrectly. Making the change to simplify_reshapes.